### ISSUE TITLE
After upgrade to 2.28.1, TextSecure crashes when opened every time.

### ISSUE THREAD

**Issue by derpston** [WhisperSystems/TextSecure#4194]
Device info: OnePlus One, Cyanogenmod 5.1.1 / 12.1, TextSecure 2.28.1
App state: Push registered, SMS enabled

I used adb logcat to capture a crash when opening TextSecure: https://gist.github.com/derpston/7f5d722a1f76d5f92116

If I disconnect the phone from the internet using airplane mode, I can open TextSecure. When I turn on wifi or 3g, TextSecure crashes as soon as it is able to make a connection. Here's a log from when that happens: https://gist.github.com/derpston/54ec33c1c8c02b65b44b

The relevant stack trace seems to be:

```java
W/AttachmentDownloadJob(10673): Downloading push part (row id: 1381, unique ID: 1443374570637)
W/JobConsumer(10673): java.lang.NegativeArraySizeException: -20
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.crypto.MasterCipher.verifyMacBody(MasterCipher.java:157)
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.crypto.MasterCipher.decryptBytes(MasterCipher.java:98)
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.crypto.MediaKey.getDecrypted(MediaKey.java:28)
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.jobs.AttachmentDownloadJob.createAttachmentPointer(AttachmentDownloadJob.java:133)
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.jobs.AttachmentDownloadJob.retrievePart(AttachmentDownloadJob.java:108)
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.jobs.AttachmentDownloadJob.onRun(AttachmentDownloadJob.java:81)
W/JobConsumer(10673): 	at org.thoughtcrime.securesms.jobs.MasterSecretJob.onRun(MasterSecretJob.java:18)
W/JobConsumer(10673): 	at org.whispersystems.jobqueue.JobConsumer.runJob(JobConsumer.java:76)
W/JobConsumer(10673): 	at org.whispersystems.jobqueue.JobConsumer.run(JobConsumer.java:46)
```

Looks like there's a message waiting that causes it to crash.

Right now I can't use TextSecure at all. Any ideas?

**Comment by derpston** [comment]
I signed up for the testing release stream with https://play.google.com/apps/testing/org.thoughtcrime.securesms and upgraded to Signal 3.0.1 - the persistent crash went away. I'd prefer to be on the stable release stream, but I guess this'll work for now.

When the crashing stopped and TS/Signal finally opened successfully, I didn't get the waiting message that the earlier logs implied I might have.

Did this verifyMacBody exception get fixed between 2.28.1 and 3.0.1?

**Comment by McLoo** [comment]
Sounds like #4061

**Comment by derpston** [comment]
Yep, this looks like the same issue, thanks! I didn't spot that issue, I didn't think to look for closed ones.

Okay, since #4061 was apparently fixed by https://github.com/WhisperSystems/TextSecure/commit/cdf982a356081329ca03b88de9d24669cbd68083 and was released in 3.0.0, and because I found that 3.0.1 fixed my problem, I'll close this issue. If I'm wrong, please reopen it. :)
